File: voice-agent/backend_agents/remote_tunnel.py
"""
RemoteTunnel - Pub/Sub tunnel for remote agent communication.

Connects to a relay server and provides:
- pub(): Publish messages to the tunnel
- sub(): Subscribe to incoming messages with a callback
"""

import asyncio
import json
import logging
import os
from typing import Callable, Optional

import websockets

from backend_agents.tunnel_interface import TunnelInterface

logger = logging.getLogger(__name__)

class RemoteTunnel(TunnelInterface):
    """Pub/Sub tunnel for remote agent communication."""

    def __init__(self, host: str = None, api_key: str = None):
        super().__init__()
        self.websocket = None
        self.agent_id = "voice-agent"
        self.host = host or os.getenv("PARTYKIT_HOST")
        self.api_key = api_key

    async def connect_to_room(
        self, room_id: str, agent_id: str = "voice-agent", role: str = "voice-agent", **kwargs
    ) -> None:
        """
        Connect to a room by room ID (convenience method).
        Handles URL formation internally.
        """
        protocol = "ws" if "localhost" in self.host else "wss"
        url = f"{protocol}://{self.host}/parties/main/{room_id}"
        return await self.connect(url, agent_id=agent_id, role=role, **kwargs)

    async def connect(
        self, url: str, agent_id: str = "voice-agent", role: str = "voice", **kwargs
    ) -> None:
        """Connect to the relay tunnel."""
        try:
            self.agent_id = agent_id
            separator = "&" if "?" in url else "?"
            api_key_param = f"&apiKey={self.api_key}" if self.api_key else ""
            full_url = f"{url}{separator}id={self.agent_id}&role={role}{api_key_param}"

            self.websocket = await websockets.connect(full_url)
            self._is_connected = True
            asyncio.create_task(self._listen())
            logger.info("✅ Connected to tunnel")
        except Exception as e:
            logger.error(f"Failed to connect: {e}")
            self._is_connected = False
            raise

    def sub(self, callback: Callable) -> None:
        """Subscribe to incoming messages with a callback."""
        self._callback = callback

    async def pub(self, message: str, agent: str = "opencode", **kwargs) -> bool:
        """
        Publish a task message to the tunnel.

        Args:
            message: The task content/description
            agent: Target agent (default: "opencode")
            **kwargs: Additional args
        """
        if not self.websocket or not self._is_connected:
            logger.error("Cannot publish: Not connected")
            return False

        payload = {
            "type": "tasks",
            "content": message,
            "agent": agent,
            "source": self.agent_id,
        }

        try:
            await self.websocket.send(json.dumps(payload))
            return True
        except Exception as e:
            logger.error(f"Failed to publish: {e}")
            return False

    async def publish(self, payload: dict) -> bool:
        """Publish a raw payload to the tunnel."""
        if not self.websocket or not self._is_connected:
            logger.error("Cannot publish: Not connected")
            return False

        try:
            await self.websocket.send(json.dumps(payload))
            return True
        except Exception as e:
            logger.error(f"Failed to publish: {e}")
            return False

    # Legacy alias
    async def send_message(self, message: str, agent: str = "opencode", **kwargs) -> bool:
        return await self.pub(message, agent=agent, **kwargs)

    async def _listen(self):
        """Listen for incoming messages and invoke subscriber callback."""
        try:
            while self._is_connected and self.websocket:
                message = await self.websocket.recv()
                await self._handle_message(message)
        except websockets.exceptions.ConnectionClosed:
            pass
        except Exception as e:
            logger.error(f"Tunnel error: {e}")
        finally:
            self._is_connected = False

    async def _handle_message(self, message: str):
        """Handle incoming messages and invoke subscriber callback."""
        try:
            data = json.loads(message)
            msg_type = data.get("type")
            logger.debug(f"📨 Received message type: {msg_type}")

            content = None
            if msg_type == "task-response":
                content = data.get("result") or data.get("error", "")
            elif msg_type == "task-update":
                status = data.get("status", "running")
                if status in ("needs_input", "completed", "error"):
                    content = data.get("message", "")
            elif msg_type == "response":
                content = data.get("content", "")
            elif msg_type == "tasks":
                # Ignore task messages (we're the sender)
                logger.debug("Ignoring tasks message (echo)")
                return

            if content and self._callback:
                if asyncio.iscoroutinefunction(self._callback):
                    await self._callback(content)
                else:
                    self._callback(content)

        except json.JSONDecodeError:
            pass
        except Exception as e:
            logger.error(f"Message error: {e}")

    async def disconnect(self) -> None:
        """Disconnect from the tunnel."""
        self._is_connected = False
        if self.websocket:
            try:
                await self.websocket.close()
            except Exception:
                pass
            self.websocket = None

File: cli/tunnel/RemoteTunnel.ts
/**
 * RemoteTunnel - Pub/Sub tunnel for remote agent communication.
 *
 * Connects to a relay server and provides:
 * - pub(): Publish messages to the tunnel
 * - sub(): Subscribe to incoming messages with a callback
 */

import {
  TunnelInterface,
  type ConnectOptions,
  type SendMessageOptions,
  type MessageCallback,
} from "./TunnelInterface";

export interface RemoteTunnelOptions {
  reconnect?: boolean;
  reconnectDelay?: number;
  host?: string;
}

const DEFAULT_HOST = "heyatlas-agents-rooms.bishwenduk029.partykit.dev";

export class RemoteTunnel extends TunnelInterface {
  private websocket: WebSocket | null = null;
  private currentUrl: string | null = null;
  private currentRoom: string | null = null;
  private options: RemoteTunnelOptions;
  private reconnectTimeout: ReturnType<typeof setTimeout> | null = null;

  constructor(options: RemoteTunnelOptions = {}) {
    super();
    this.options = {
      reconnect: true,
      reconnectDelay: 3000,
      ...options,
    };
  }

  /**
   * Connect to a room by room ID (convenience method).
   * Handles URL formation internally.
   */
  async connectToRoom(
    roomId: string,
    options: ConnectOptions = {},
  ): Promise<void> {
    const host = this.options.host || DEFAULT_HOST;
    const protocol = host.includes("localhost") ? "ws" : "wss";
    const url = `${protocol}://${host}/parties/main/${roomId}`;
    return this.connect(url, options);
  }

  /**
   * Connect to a relay tunnel.
   */
  async connect(url: string, options: ConnectOptions = {}): Promise<void> {
    try {
      this._agentId = options.agentId || this._agentId;
      const role = options.role || "agent";

      const separator = url.includes("?") ? "&" : "?";
      const fullUrl = `${url}${separator}id=${this._agentId}&role=${role}`;
      this.currentUrl = url;

      // Extract room name from URL
      const roomMatch = url.match(/\/parties\/\w+\/([^?]+)/);
      this.currentRoom = roomMatch?.[1] ?? null;

      this.websocket = new WebSocket(fullUrl);

      await new Promise<void>((resolve, reject) => {
        if (!this.websocket) return reject(new Error("WebSocket not created"));

        this.websocket.onopen = () => {
          this._isConnected = true;
          resolve();
        };

        this.websocket.onerror = (error) => {
          reject(error);
        };

        this.websocket.onclose = () => this.handleClose();
        this.websocket.onmessage = (event) => this.handleMessage(event);
      });
    } catch (error) {
      this._isConnected = false;
      throw error;
    }
  }

  /**
   * Publish a task message to the tunnel.
   */
  async pub(
    message: string,
    options: SendMessageOptions = {},
  ): Promise<boolean> {
    if (!this.websocket || !this._isConnected) {
      return false;
    }

    const payload = {
      type: "tasks",
      content: message,
      agent: options.agent || "opencode",
      source: this._agentId,
      ...options,
    };

    try {
      this.websocket.send(JSON.stringify(payload));
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Publish a raw payload to the tunnel.
   */
  async publish(payload: Record<string, unknown>): Promise<boolean> {
    if (!this.websocket || !this._isConnected) {
      return false;
    }

    try {
      this.websocket.send(JSON.stringify(payload));
      return true;
    } catch {
      return false;
    }
  }

  // Legacy aliases
  async sendMessage(
    message: string,
    options: SendMessageOptions = {},
  ): Promise<boolean> {
    return this.pub(message, options);
  }

  async send(payload: Record<string, unknown>): Promise<boolean> {
    return this.publish(payload);
  }

  /**
   * Disconnect from the tunnel.
   */
  async disconnect(): Promise<void> {
    this._isConnected = false;
    this.currentUrl = null;
    this.currentRoom = null;

    if (this.reconnectTimeout) {
      clearTimeout(this.reconnectTimeout);
      this.reconnectTimeout = null;
    }

    if (this.websocket) {
      try {
        this.websocket.close();
      } catch {
        // ignore
      }
      this.websocket = null;
    }
  }

  /**
   * Get the current room name.
   */
  get room(): string | null {
    return this.currentRoom;
  }

  private handleClose(): void {
    const wasConnected = this._isConnected;
    this._isConnected = false;

    if (this.options.reconnect && this.currentUrl && wasConnected) {
      this.reconnectTimeout = setTimeout(() => {
        if (this.currentUrl) {
          this.connect(this.currentUrl, { agentId: this._agentId }).catch(
            () => {},
          );
        }
      }, this.options.reconnectDelay);
    }
  }

  /**
   * Handle incoming messages and invoke subscriber callback.
   */
  private async handleMessage(event: MessageEvent): Promise<void> {
    try {
      const data = JSON.parse(event.data.toString());
      const msgType = data.type;

      let content: string | null = null;

      if (msgType === "task-response") {
        content = data.result || data.error;
      } else if (msgType === "task-update") {
        const status = data.status;
        if (["needs_input", "completed", "error"].includes(status)) {
          content = data.message || "";
        }
      } else if (msgType === "response") {
        content = data.content || "";
      } else if (msgType === "tasks") {
        content = data.content;
      }

      if (content && this._callback) {
        await Promise.resolve(this._callback(content, data));
      }
    } catch {
      // Silently ignore parse errors
    }
  }
}
